extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -167,12 +183,36 @@ public SecurityIdentity apply(Throwable throwable) {
             }
         }
 
-        challenge = new ChallengeData(HttpResponseStatus.FOUND.code(), HttpHeaders.LOCATION,
-                configContext.auth.authorizeURL(params));
+        if (shouldAutoRedirect(configContext, context)) {
+            challenge = new ChallengeData(HttpResponseStatus.FOUND.code(), HttpHeaders.LOCATION,
+                    configContext.auth.authorizeURL(params));
+        } else {
+            // If the client (usually an SPA) wants to handle the redirect manually, then
+            // return status code 399 and in the www-authenticate header put the redirect url
+            // www-authenticate header content; No standard content matched the exact use case
+            // so the OAuth redirect-url= is an arbitrary string that can be tested
+            // for on the client side
+            String headerContent = "OAuth redirect-url=" + configContext.auth.authorizeURL(params);
+            // Vertx HTTPHeaders does noy have a "WWW_AUTHENTICATE" static
+            // So, header name is hard coded for the moment
+            challenge = new ChallengeData(399, "WWW-Authenticate", headerContent);
+        }
 
         return Uni.createFrom().item(challenge);
     }
 
+    private boolean isXHR(RoutingContext context) {
+        return "XMLHttpRequest".equals(context.request().getHeader("X-Requested-With"));
+    }
+
+    private boolean isAutoRedirectSet(TenantConfigContext configContext) {
+        return configContext.oidcConfig.authentication.autoRedirect;
+    }
+
+    private boolean shouldAutoRedirect(TenantConfigContext configContext, RoutingContext context) {
+        return isXHR(context) ? isAutoRedirectSet(configContext) : true;
+    }
+
```

Comment:
I messed up something here. This is duplicate code. The version at 140 - 155 is the latest. I obviously didn't quite do the "update a PR" incantation correctly.